### PR TITLE
Improve relational model integration in Query

### LINKED ISSUE #32853: Clean up default mappings
Currently, `entityType.GetDefaultMappings()` should only be used in a `FromSql` context. Elsewhere, if the entity type is not mapped to a table, a view, a SqlQuery or a function, an exception should be thrown.

Also, add more tests (particularly negative cases) for `FromSql` and `ToSqlQuery`.

### LINKED ISSUE #35466: Query: audit uses of entityType.GetViewOrTableMappings().Single (OrDefault) in the query pipeline.
entityType.GetViewOrTableMappings().Single() / SingleOrDefault() is a pattern that we use in several places in our code. It can result in sequence contains more than one element in the entity splitting scenario (see https://github.com/dotnet/efcore/issues/35293). We should make sure all the places where we use this pattern are correct.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #38060
Fixes #35466
Fixes #32853

The query pipeline previously fell back to model‑wide accessors (`GetJsonPropertyName`, container‑column‑by‑name scans) at translation time, which produced wrong/ambiguous results under entity splitting, TPT, and JSON‑on‑view scenarios. This change threads the actual table set and JSON element metadata through the existing relational query expressions and consolidates priority logic in one place.

## Changes

- **New `GetQueryMappings()` internal helper** (RelationalTypeBaseExtensions) returns storage mappings in priority order — *default* SQL query → *default* function → view → table — and replaces all `GetViewOrTableMappings()` call sites in the query pipeline. Casts directly to the concrete `List<TMapping>` runtime type and filters in place to avoid LINQ allocations.
- **`StructuralTypeProjectionExpression.TableMap`** (optional `IReadOnlyDictionary<ITableBase, string>`) — threaded through TPT, single‑table, entity‑splitting and TPC `CreateSelect` branches, preserved across `MakeNullable`, `UpdateEntityType`, `VisitChildren`, set‑op merge, and subquery lift. `BindProperty`, `TryRewriteEntityEquality`, and `GenerateMaterializationCondition` now scope their principal‑split‑table lookup to the actually‑projected tables.
- **`JsonQueryExpression.FindJsonElement(IPropertyBase)`** — returns the `IRelationalJsonElement` whose containing column matches `JsonColumn` when available, otherwise the first element from the property's mappings. `BindProperty`/`BindStructuralProperty` and the OPENJSON/json_each translators in `RelationalQueryableMethodTranslatingExpressionVisitor`, `SqlServer`, and `Sqlite` now use this instead of model‑wide `GetJsonPropertyName()`.
- **`JsonProjectionInfo.JsonColumn`** + **`ColumnExpression.Column` preservation** in `MakeNullable`/`ApplyTypeMapping` — the shaper now uses the column resolved at CreateSelect time, falling back to a model scan only when there's no underlying `IColumnBase` (synthetic JSON columns over OPENJSON/json_each).

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.CreateSelect.cs [hunk 1]
```diff
@@ -303,12 +303,15 @@ Expression ProcessComplexPropertyTpc(
...
-                            // same name)
+                            // same name).
+                            // The per-table IColumnBase flows through the inner projection so the outer union column carries a
+                            // model-column reference.
                             projections.Add(
                                 new ProjectionExpression(
                                     new ColumnExpression(
                                         column.Name,
                                         tableAlias,
+                                        column,
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.CreateSelect.cs]: 👀

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -4283,7 +4248,14 @@ private static ColumnExpression CreateColumnExpression(ProjectionExpression subq
...
             tableAlias,
-            column: subqueryProjection.Expression is ColumnExpression { Column: { } column } ? column : null,
+            // Preserve the underlying model column through subquery boundaries so JsonQueryExpression.FindJsonElement keeps
+            // matching after lifts (e.g. LiftJsonQueryFromSubquery wraps the json column in a JsonScalarExpression).
+            column: subqueryProjection.Expression switch
+            {
+                ColumnExpression { Column: { } column } => column,
+                JsonScalarExpression { Json: ColumnExpression { Column: { } jsonColumn } } => jsonColumn,
+                _ => null
+            },
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs]: 👀

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 3]
```diff
@@ -1286,10 +1286,13 @@ Expression CopyProjectionToOuter(SelectExpression innerSelectExpression, Express
...
                             }
 
+                            // The outer projection still references the same underlying JSON column; only the
+                            // index into the outer projection changes, so preserve JsonColumn for the shaper.
                             remappedConstant = Constant(
                                 new JsonProjectionInfo(
                                     projectionIndexMap[jsonProjectionInfo.JsonColumnIndex],
-                                    newKeyAccessInfo));
+                                    newKeyAccessInfo,
+                                    jsonProjectionInfo.JsonColumn));
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1296]: 👀

File: src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -555,10 +556,11 @@ IEntityType entityType
                     .Where(n => n.ForeignKey.IsOwnership
                         && n.TargetEntityType.IsMappedToJson()
                         && n.ForeignKey.PrincipalToDependent == n)
-                    .Select(n => n.TargetEntityType.GetJsonPropertyName() ?? throw new UnreachableException()),
+                    .Select(n => jsonQueryExpression.GetJsonElement(n).PropertyName ?? throw new UnreachableException()),
 
             IComplexType complexType
-                => complexType.GetComplexProperties().Select(p => p.ComplexType.GetJsonPropertyName() ?? throw new UnreachableException()),
+                => complexType.GetComplexProperties()
+                    .Select(p => jsonQueryExpression.GetJsonElement(p).PropertyName ?? throw new UnreachableException()),
```
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Query/Internal/SqlServerQueryableMethodTranslatingExpressionVisitor.cs]: Should add some placeholder text to these exceptions

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs]: This looks a bit odd... The idea here was for ColumnExpression to reference the relational model IColumn it refers to, similar to how TableExpression references its the relational model ITableBase; IIRC I did this specifically so that e.g. the translation can look at a ColumnExpression and know whether it has a database index, since that can affect the translation strategy.

This change makes it so that a ColumnExpression referring to a JsonScalarExpression will point the the relational model column which **contains** that JSON scalar, which doesn't seem right (or at least changes the meaning of ColumnExpression.Column significantly). I'd propose leaving this null at least for now - ColumnExpression.Column is optional anyway, and doesn't have to be populated.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:1296]: Makes sense to me.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.CreateSelect.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.CreateSelect.cs]: Hmm... This is the code that creates all the UNION subqueries for a TPC hierarchy; here we're creating the projection for a single such subquery. Linking the subquery's projected column to the relational model seems correct; but the comment just above seems wrong: the column projected by the outer query (the UNION) cannot (I think) be said to reference any specific column in the relational model, since it's the result of a UNION between several.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs]: Ok, changed this to only set the containing column when the result is going to be JsonQueryExpression.JsonColumn to preserve the semantics

**Review by cincuranet — APPROVED** [review]
